### DISCUSSION TITLE
Error while compiling EDIPIC-2D

### DISCUSSION THREAD

**Discussion by RemoYuan** [PrincetonUniversity/EDIPIC-2D#18] [Q&A] [answered]
Hi, I have installed the MPI, PETSc and HYPRE successfully, but received a error when I tried to compiling the EDIPIC-2D with the code bellow:

export PETSC_DIR=path_to_petsc_installation
export LD_LIBRARY_PATH=${PETSC_DIR}/lib:${LD_LIBRARY_PATH}
cd src
make

The error is as followings. The petsc folder is under the EDIPIC-2D/

gfortran
"Using gcc/gfortran compilers"
mpif90 -O2 -ffree-line-length-none -Ipath_to_petsc_installation/include -c pic2d_PETSc_Solver.F90
f951: Warning: Nonexistent include directory ‘path_to_petsc_installation/include’ [-Wmissing-include-dirs]
pic2d_PETSc_Solver.F90:15:2:

   15 |
      |  1
Fatal Error: petsc/finclude/petsc.h: No such file or directory
compilation terminated.
make: *** [Makefile:93: pic2d_PETSc_Solver.o] Error 1
![error](https://github.com/PrincetonUniversity/EDIPIC-2D/assets/170242384/f7214d5b-d244-4443-990b-c5b5744abe1c)

**Comment by stephethier** [comment]
Hi Remo,
You need to replace the string "path_to_petsc_installation" with the actual path where you installed PETSc so that the compiler can find the petsc.h file and the libraries "libpetsc.a" and "libHYPRE.a". Did you build PETSc and HYPRE from the source code or via a software package manager? 

cheers,
Stephane Ethier

**Reply by RemoYuan** [reply]
Hi there,
Thanks for your help, the errors are solved by replacing the string "path_to_petsc_installation" with the actrual path. The compiling is nearly down. However, there comes some new errors and warnings, which can be seen as follows:  
![image](https://github.com/PrincetonUniversity/EDIPIC-2D/assets/170242384/41ba7b75-f3bd-426a-9363-664628ccb710)

**Comment by stephethier** [comment]
Hi Yuan,
The first thing you should probably do is to rename the directory where you installed the PETSc include files and libraries. You named it `*path_to_final_installation*` but the star symbol (`*`) under Linux is a wildcard so it shouldn't be used for file or directory names. I normally use the name of the compiler and MPI distribution, sometimes with the version number. Something like `gcc12_mpich`. So in your case it would end up being `/home/yuanchg/petsc/gcc12_mpich/`.

As for the error that you get, can you please go to the location where the PETSc library`libpetsc.a` is located (`/home/yuanchg/petsc/*path_to_final_installation*/lib`) and run the following command:

`nm libpetsc.a | grep petscinitialize
`
and send me the result.
Thanks.

**Reply by RemoYuan** [reply]
Hi, Ethier
The result of the command is as following:
![image](https://github.com/PrincetonUniversity/EDIPIC-2D/assets/170242384/f8b96390-16b1-4466-95be-7fac6ebf7ca2)
I am sorry I have deleted the direction(/home/yuanchg/petsc) and have to run the command in a backup of the petsc direction in /EDIPIC-2D. I hope it works.

**Comment by stephethier** [comment]
Ok. It looks like the installation only generated the shared libraries. Can you see a file named `libpetsc.so`? If yes, then try the same command on that library:

`nm libpetsc.so | grep petscinitialize`

**Reply by RemoYuan** [reply]
It seems there isn't a file named libpetsc.so in the direction /*path_to_final_installation*/lib, but there are some at other diections. 
![image](https://github.com/PrincetonUniversity/EDIPIC-2D/assets/170242384/ccd1a3db-505a-43ba-81fb-6f26683f34a3)

**Comment by stephethier** [comment, marked as answer]
Ok. Maybe you forgot to run `make install` at the end of the PETSc installation. Try the following:
```
export PETSC_DIR=/home/yuanchg/EDIPIC-2D/petsc/compiler_name_and_version
cd /home/yuanchg/EDIPIC-2D/src
make clean
make
```

**Reply by RemoYuan** [reply]
Thank you very much for answering my question, and the problem has been solved after I tried compiling the PETSc again. Now a file named edipic2d have been created and I can go on learning the EDIPIC. 
Best Wish!
Cheng Yuan.

**Reply by stephethier** [reply]
Great! Glad that it is working now. 
Don't hesitate to contact us if you have other questions.
